Add a crew monitoring server

juliangiebel opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->
* Adds crew monitoring servers
  * Only one is active at a time. If the one that is currently active loses power or gets removed / destroyed another available server will take over
* Changes suit sensors to not broadcast anymore
* Adds a random offset to each suit sensors update time

Suit sensors don't just broadcast their status anymore! 
Instead they retrieve the currently active crew monitoring servers address from the crew monitoring system and send their status directly to that server.

That server is then periodically broadcasting the status of all "connected" suit sensors in one packet on a different frequency.

So we went from a huge amount of broadcasting devices to just one.
  
**Changelog**
<!--
Here you can fill out a changelog that will automatically be added to the game when your PR is merged
There are 4 icons for changelog entries: add, remove, tweak, fix. I trust you can figure out the rest.

You can put your name after the :cl: symbol to change the name that shows in the changelog (otherwise it takes your GitHub username)
Like so: :cl: PJB

Generally, only put things in changelogs that players actually care about. Stuff like "Refactored X system, no changes should be visible" shouldn't be on a changelog.

For writing actual entries, don't consider the entry type suffix (e.g. add) to be "part" of the sentence:
bad: - add: a new tool for engineers
good: - add: added a new tool for engineers
-->

:cl:
- add: The crew monitoring server. It relays the status of suit sensors to crew monitors.

moonheart08 commented:
Does this work when multiple stations are loaded.

juliangiebel commented:
It doesn't break but you still see all crew members regardles of what station they are on.

I could restrict all of the monitoring equipment to only send and receive only for the grid they are on and have one active server per grid.
I would only need to figure out how to make sure suit sensors work of station too.
A frequency offset might work but that will break with to many stations.

Fishfish458 commented:
> 

I assume the ultimate goal would be distance based from server but i don't think that is cheap to process currently.

Emisse commented:
would there be a way to tie crew to the station grid they join in on

Macoron commented:
I understand this solution for performance reasons, but have some feedback gameplay-wise:

- When you leave 500 tiles range from a server, your suit sensor won't work. And even if you build another one on your grid it won't help, because all crew monitorings are connected to one world server. So if you have two stations with 500 tiles distance between them - one station won't be able to use their crew monitoring devices. 
- There is no way to build a server. You probably need construction graph and circuit.
- Mappers should map it asap after this PR merge.
- Crew monitoring devices should show something like "Can't connect to the server..." in UI when they can't hear from the server for a long time

**Heresy idea:** Maybe don't use device net for suit sensors and replace them just by C# function call? C# function calls should be much cheaper than packet serialize/deserialize +raising events.

metalgearsloth commented on Content.Server/Medical/CrewMonitoring/CrewMonitoringServerSystem.cs:
```diff
@@ -0,0 +1,169 @@
+            return;
+        _updateDif = 0f;
+
+        if (_activeServer == null)
+        {
+            var servers = EntityManager.EntityQuery<CrewMonitoringServerComponent>().GetEnumerator();
+            CrewMonitoringServerComponent? server = null;
+
+            //Retrieve the first available server
+            while (servers.MoveNext())
+            {
+                if (servers.Current.Available)
+                {
+                    server = servers.Current;
+                }
+            }
+
+            servers.Dispose();
+
+            if (server == null)
+                return;
+
+            _activeServer = server.Owner;
+            _deviceNetworkSystem.ConnectDevice(server.Owner);
+        }
+
+        if (_activeServer == null)
+            return;
```
Making this a singleton seems bad.

juliangiebel replied:
I'll most likely make this one server per station/grid so it wouldn't be a singleton anymore at that point.

juliangiebel commented:
> I understand this solution for performance reasons, but have some feedback gameplay-wise:
> 
> * When you leave 500 tiles range from a server, your suit sensor won't work. And even if you build another one on your grid it won't help, because all crew monitorings are connected to one world server. So if you have two stations with 500 tiles distance between them - one station won't be able to use their crew monitoring devices.
> 
> * There is no way to build a server. You probably need construction graph and circuit.
> 
> * Mappers should map it asap after this PR merge.
> 
> * Crew monitoring devices should show something like "Can't connect to the server..." in UI when they can't hear from the server for a long time
> 
> 
> **Heresy idea:** Maybe don't use device net for suit sensors and replace them just by C# function call? C# function calls should be much cheaper than packet serialize/deserialize +raising events.

I'll make this one server per station/grid. Having a range makes some sense then but I should propably increase the range.

I forgot to add construction. I will add construction and the "Can't connect to the server.." text after I get back from work today.

juliangiebel commented:
> packet serialize/deserialize

It's not serialization. You don't even *need* to split your data into seperate fields in the packets payload, you can just slab your data in as is. (Look at how CrewMonitoringServerSystem.cs) does it.
It's just that if it's not a primitive type or it doesn't have some custom logic players won't be able to see the contents of that field.

In the case of suit sensors it's about reusable code. If I write a check for what station the sender and receiver is on you can use it for other devices too instead of having that in the entity system of each of the monitoring devices.

The suit sensor code for creating a device net packet is a bit missleading tbh. Especially with it calling the process serialization/deserialization.

I can even give crew monitoring it's own device net id so incrementing the frequency for each new station would be possible instad of a check if receiver and sender are on the same grid. At that point I can also change the code that creates the suit sensors packet to just put the data in one field as players wil be unnable to see that packet anyway.

juliangiebel commented:
This is ready for review

metalgearsloth commented on Content.Server/DeviceNetwork/Components/StationLimitedNetworkComponent.cs:
```diff
@@ -0,0 +1,16 @@
+        /// Uses the grid id until moonys station beacon system is implemented
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        public GridId? StationId;
```
EntityUid

metalgearsloth commented on Content.Server/Medical/CrewMonitoring/CrewMonitoringServerComponent.cs:
```diff
@@ -0,0 +1,32 @@
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite)]
+    public bool Available = true;
+
+    /// <summary>
+    /// The id of the station the server is on.
+    /// </summary>
+    public GridId? StationId;
```
Shouldn't need to cache this.

metalgearsloth commented on Content.Server/Medical/CrewMonitoring/CrewMonitoringServerSystem.cs:
```diff
@@ -0,0 +1,167 @@
+        _updateDif += frameTime;
+        if (_updateDif < UpdateRate)
+            return;
+        _updateDif = 0f;
```
```suggestion
        _updateDif -= UpdateRate;
```

metalgearsloth reviewed (changes requested):
Replace GridId with EntityUids.

moonheart08 reviewed (dismissed):
StationSystem's in, while the beacons are missing they're not necessary for this PR's functionality. So all GridId needs changed up.

PaulRitter reviewed (dismissed):
partial review i apparently never submitted so here you go

juliangiebel commented on Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs:
```diff
@@ -50,18 +50,24 @@ public override void Update(float frameTime)
         /// <param name="data">The data to be sent</param>
-        public void QueuePacket(EntityUid uid, string? address, NetworkPayload data, uint? frequency = null, DeviceNetworkComponent? device = null)
+        /// <returns>Returns true when the packet was successfully enqueued.</returns>
+        public bool QueuePacket(EntityUid uid, string? address, NetworkPayload data, uint? frequency = null, DeviceNetworkComponent? device = null)
```
If frequency is null it get's set to the frequency the device network component has.
If the address is null the package get broadcasted. Both can be null at the same time.

PaulRitter replied:
makes sense

juliangiebel commented:
I addressed the reviews and made it use station system.
Now There can be one active server per station and the suit sensors get assigned to a station on component init.

juliangiebel commented:
This is ready to be merged again

ElectroJr reviewed (dismissed):
Just some minor guff, otherwise seems good to merge. 

Though for whoever hits the button: they should ping mappers because this will effectively break sensors if maps don't get updated before the next publish.

juliangiebel commented:
I finally got around to fixing up this PR.

mirrorcult commented on Content.Server/DeviceNetwork/Systems/StationLimitedNetworkSystem.cs:
```diff
@@ -0,0 +1,68 @@
+    [UsedImplicitly]
+    public sealed class StationLimitedNetworkSystem : EntitySystem
+    {
+        [Dependency] private readonly StationSystem? _stationSystem = default!;
```
```suggestion
        [Dependency] private readonly StationSystem _stationSystem = default!;
```
will always be valid for entity systems

mirrorcult commented on Content.Server/DeviceNetwork/Systems/StationLimitedNetworkSystem.cs:
```diff
@@ -0,0 +1,68 @@
+        /// </summary>
+        private void OnMapInit(EntityUid uid, StationLimitedNetworkComponent networkComponent, MapInitEvent args)
+        {
+            networkComponent.StationId = _stationSystem?.GetOwningStation(uid);
```
```suggestion
            networkComponent.StationId = _stationSystem.GetOwningStation(uid);
```

mirrorcult commented on Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:
```diff
@@ -93,6 +117,11 @@ private void OnMapInit(EntityUid uid, SuitSensorComponent component, MapInitEven
             }
         }
 
+        private void OnComponentInit(EntityUid uid, SuitSensorComponent component, ComponentInit args)
+        {
+            component.StationId = _stationSystem.GetOwningStation(uid);
+        }
```
why not on mapinit as well?

juliangiebel replied:
I moved it to mapinit and added a data field annotation

mirrorcult commented on Resources/Prototypes/Entities/Structures/Machines/crew_monitor_server.yml:
```diff
@@ -0,0 +1,57 @@
+  id: CrewMonitoringServer
+  parent: BaseMachinePowered
+  name: crew monitoring server
+  description: Receives and relais the status of all active suit sensors on the station
```
```suggestion
  description: Receives and relays the status of all active suit sensors on the station.
```

mirrorcult reviewed (dismissed):
this will need to be mapped into the bridge correct?

juliangiebel commented:
@mirrorcult I implemented your suggestions
